fhir-persistence-schema/docs/DB2MultiTenancy.md

```diff
@@ -1,16 +1,225 @@
+
+Logically:
+
+```
+SET ptng.session_tenant = 1;
+SELECT * FROM ptng.patients;
+commit;
+```
+
+Internally, DB2 rewrites the query to look like this:
+
+```
+SELECT * FROM ptng.patients WHERE mt_id = 1;
+```
+
+Because the table `ptng.patients` is partitioned using mt_id, the query optimizer is able to perform a partition-prune, by limiting the query access to the partition associated with the tenant. This results in a query execution plan very similar in terms of performance to a single-tenant scenario without any partitioning.
+
+To isolate a tenant's data at the file-system level, a tablespace is created per tenant and the partition objects are are allocated to that tablespace, including any local indexes:
+
+```
+CREATE TABLESPACE TS_TENANT1 MANAGED BY AUTOMATIC STORAGE;
+CREATE TABLESPACE TS_TENANT2 MANAGED BY AUTOMATIC STORAGE;
+```
+
+
+Adding new tenants requires adding partitions to each table in the data schema. Note how the partition is configured with the tenant's specific tablespace:
+
+```
+-- for tenant #1
+ALTER TABLE ptng.patients ADD PARTITION TENANT1 STARTING 1 INCLUSIVE ENDING 1 INCLUSIVE IN TS_TENANT1;
+
+-- for tenant #2
+ALTER TABLE ptng.patients ADD PARTITION TENANT2 STARTING 2 INCLUSIVE ENDING 2 INCLUSIVE IN TS_TENANT2;
+```
+
+When the tenant is deleted, the partition associated with the tenant is detached and subsequently dropped from the database. This renders the tenant data inaccessible from the perspective of a database user. However, database space management does not necessarily overwrite the freed blocks. Although the data is encrypted, someone with the database key and sufficient time, tooling and knowledge could theoretically retrieve parts of the data. By using a tablespace which is dropped, the database will delete the files at the OS layer.
+
+The process to delete a tenant is (logically):
+
+```
+ALTER TABLE <tbl> DETACH PARTITION INTO <tbl_drp1>;
+...
+COMMIT;
+DROP TABLE <tbl_drp1>;
+...
+COMMIT;
+DROP TABLESPACE TS_TENANT1;
+```
+
+In this case, TS_TENANT1 is the name of the tablespace previously assigned to the tenant being dropped.
+
+Note that DB2 requires a commit after the partition is detached and before the table `<tbl_drp1>` can be dropped. The tooling must implement this process to be idempotent to handle cases where a detach or drop might fail.
+
+Dropping partitions in this way works by swizzling metadata in the catalog which requires very little (redo) logging and is therefore far more efficient than deleting actual rows of data.
+
+
+
+
+## Tenant Authentication
+
+The server connect to Db2 with a user configured for the least privileges it requires to provide the FHIR resource persistence and search functions. This user does not have any database administration privileges, unlike the FHIRADMIN user which is used to deploy the schema, allocate tenants etc.
+
+![Privilege Design](privilege_design.png)
```

Comment:
This image needs some work. 

Schema is FHIR_ADMIN
I think the red lines should be removed
the CREATE PERMISSIONS should be removed at the bottom